CSLabs.Api/Proxmox/ProxmoxApi.cs

```diff
@@ -40,6 +42,49 @@ private async Task Login()
             _loggedInAt = DateTime.Now;
         }
 
+        public async Task ManageApiToken()
+        {
+            await LoginIfNotLoggedIn();
+            var userid = $"{HypervisorNode.Hypervisor.UserName}@pam";
+            if (string.IsNullOrEmpty(client.ApiToken))
+                await GenerateApiToken(userid);
+            else
+                await RotateApiToken(userid);
+        }
+
+        private async Task RotateApiToken(string userid)
+        {
+            await LoginIfNotLoggedIn();
+            await PerformRequest(() => client.Access.Users[userid].Token["API_TOKEN"].RemoveToken());
+            await GenerateApiToken(userid);
+        }
+
+        private async Task GenerateApiToken(string userid)
+        {
+            await LoginIfNotLoggedIn();
+            var apiTokenResponse = await GetApiToken(userid);
+            client.ApiToken = $"{userid}!API_TOKEN={apiTokenResponse.Value}";
+        }
+
+        private async Task<ApiTokenResponse> GetApiToken(string userid)
+        {
+            await LoginIfNotLoggedIn();
+            var expireDate = (int) DateTimeOffset.UtcNow.AddDays(30).ToUnixTimeSeconds();
+            var apiKey = await PerformRequest(() =>
+                client.Access.Users[userid].Token["API_TOKEN"].GenerateToken(expire: expireDate, privsep: true));
+
+            return new ApiTokenResponse()
+            {
+                FullTokenId = ((IDictionary<string, object>) apiKey.Response.data)["full-tokenid"].ToString(),
+                Info = new TokenInfo()
+                {
+                    Expire = int.Parse(apiKey.Response.data.info.expire),
+                    PrivSep = apiKey.Response.data.info.privsep == "1"
+                },
+                Value = apiKey.Response.data.value
+            };
+        }
+
         public async Task<TicketResponse> GetTicket(int vmId)
```

Comment:
@jasekiw Using API tokens gets rid of the need to log in with username and proxmox. The Login method we use logs in and generates a ticket, which is used to authenticate noVNC in cslabs-webapp/src/api/rfb.ts. Since API tokens will remove the need for log in, a ticket will not be generated to authenticate to noVNC, and I am receiving a 401 unauthorized. Reference to this proxmox API ![here](https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu/{vmid}/vncproxy). I may have missed something, I am not sure what we can do or maybe there is another way?

Reply:
Is there a time you are available to go through this via zoom? I would like you to demonstrate the issue for me if you can.

Reply:
Yes, evening times work best. Tomorrow after the meeting?